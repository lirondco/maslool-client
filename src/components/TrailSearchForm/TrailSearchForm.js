import React, { Component } from 'react'
import './TrailSearchForm.css'
import AllTrailsContext from '../../contexts/AllTrailsContext'
import TrailsApiService from '../../services/trails-api-service'
import { Input, Label, Button } from '../Utils/Utils'

export default class TrailSearchForm extends Component {
    static defaultProps = {
        onSearchSuccess: () => { }
    }

    static contextType = AllTrailsContext

    firstInput = React.createRef()


    handleSearch = ev => {
        this.context.setError({ error: 'Processing your request...'})
        ev.preventDefault()
        this.context.clearSearchResult()
        const { search, location, difficulty, rating } = ev.target
        TrailsApiService.searchTrails(`?name=${search.value}&region=${location.value}&difficulty=${difficulty.value}&rating=${rating.value}`)
            .then(this.context.setSearchResult)
            .then(this.props.onSearchSuccess)
            .then(this.context.clearError)
            .catch(this.context.setError)
    }

    componentDidMount = () => {
        this.firstInput.current.focus()
        this.context.clearError()
        TrailsApiService.searchTrails('')
            .then(this.context.setAllTrails)
            .catch(this.context.setError)
    }

    render() {
        const { allTrails } = this.context
        let uniqueLocations = [...new Set(allTrails.map(trail => trail.location.region))]
        const difficulties = ['Beginner', 'Intermediate', 'Advanced']
        const ratings = [1, 2, 3, 4, 5]

        return (
            <div className='TrailSearchForm'>
                <h3>Search for your next hike</h3>
                <form className='search_form' onSubmit={this.handleSearch}>
                    <Label htmlFor='search_box'>
                        Search by name:
                    </Label>
                    <Input
                        ref={this.firstInput}
                        id='search_box'
                        name='search'
                    />
                    <Label htmlFor='location_search'>
                        Search by location:
                    </Label>
                    <select id='location_search' name='location'>
                        <option value={''}>Any</option>
                        {uniqueLocations.map(location => {
                            return (
                                <option key={location.toLowerCase()} value={location.toLowerCase()}>
                                    {location}
                                </option>
                            )
                        })}
                    </select>
                    <Label htmlFor='difficulty_search'>
                        Search by difficulty:
                    </Label>
                    <select id='difficulty_search' name='difficulty'>
                        <option value={''}>Any</option>
                        {difficulties.map(diff => {
                            return (
                                <option key={diff.toLowerCase()} value={diff.toLowerCase()}>
                                    {diff}
                                </option>
                            )
                        })}
                    </select>
                    <Label htmlFor='rating_saerch'>
                        Search by minimum rating:
                    </Label>
                    <select id='rating_search' name='rating'>
                        <option value={0}>Any</option>
                        {ratings.map(r => {
                            return (
                                <option key={r} value={Number(r)}>
                                    {r}
                                </option>
                            )   
                        })}
                    </select>
                    <p className='leave_fields_text'>Leave fields blank to return all.</p>
                    <Button type='submit'>
                        Search
                    </Button>
                    <div role='alert'>
                        {this.context.error && <p className='error'>{this.context.error.error}</p>}
                    </div>
                </form>
            </div>
        )
    }

}