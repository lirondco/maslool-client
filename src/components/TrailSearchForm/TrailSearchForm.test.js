import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { withContext } from "shallow-with-context";
import TrailSearchForm from "./TrailSearchForm";

describe(`TrailSearchForm component`, () => {
  const context = {
    allTrails: [{id: 1, location: { region: 'Pennsylvania' } }],
    clearError: () => {},
    setError: () => {},
    clearSearchResult: () => {}
  };
  const ComponentWithContext = withContext(TrailSearchForm, context);

  it("renders the form", () => {
    const wrapper = shallow(<ComponentWithContext />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("form submits when submitted", () => {
    const wrapper = shallow(<ComponentWithContext />, { context });
    const mockedEvent = {
      preventDefault: () => {},
      target: {
        search: { value: "name" },
        difficulty: { value: "Beginner" },
        rating: { value: 3 },
        location: { value: 'Pennsylvania' }
      },
    };
    wrapper.find("form").at(0).simulate("submit", mockedEvent);
  });
});
