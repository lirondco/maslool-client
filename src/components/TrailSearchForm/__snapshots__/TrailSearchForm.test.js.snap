// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrailSearchForm component renders the form 1`] = `
<div
  className="TrailSearchForm"
>
  <h3>
    Search for your next hike
  </h3>
  <form
    className="search_form"
    onSubmit={[Function]}
  >
    <Label
      htmlFor="search_box"
    >
      Search by name:
    </Label>
    <ForwardRef
      id="search_box"
      name="search"
    />
    <Label
      htmlFor="location_search"
    >
      Search by location:
    </Label>
    <select
      id="location_search"
      name="location"
    >
      <option
        value=""
      >
        Any
      </option>
      <option
        key="pennsylvania"
        value="pennsylvania"
      >
        Pennsylvania
      </option>
    </select>
    <Label
      htmlFor="difficulty_search"
    >
      Search by difficulty:
    </Label>
    <select
      id="difficulty_search"
      name="difficulty"
    >
      <option
        value=""
      >
        Any
      </option>
      <option
        key="beginner"
        value="beginner"
      >
        Beginner
      </option>
      <option
        key="intermediate"
        value="intermediate"
      >
        Intermediate
      </option>
      <option
        key="advanced"
        value="advanced"
      >
        Advanced
      </option>
    </select>
    <Label
      htmlFor="rating_saerch"
    >
      Search by minimum rating:
    </Label>
    <select
      id="rating_search"
      name="rating"
    >
      <option
        value={0}
      >
        Any
      </option>
      <option
        key="1"
        value={1}
      >
        1
      </option>
      <option
        key="2"
        value={2}
      >
        2
      </option>
      <option
        key="3"
        value={3}
      >
        3
      </option>
      <option
        key="4"
        value={4}
      >
        4
      </option>
      <option
        key="5"
        value={5}
      >
        5
      </option>
    </select>
    <p
      className="leave_fields_text"
    >
      Leave fields blank to return all.
    </p>
    <ForwardRef
      type="submit"
    >
      Search
    </ForwardRef>
    <div
      role="alert"
    />
  </form>
</div>
`;
